url/src/main/java/com/datastax/oss/dsbulk/url/S3URLStreamHandler.java

```diff
@@ -96,20 +100,25 @@ InputStream getInputStream(S3Client s3Client, GetObjectRequest getObjectRequest)
 
     @VisibleForTesting
     S3Client getS3Client(String query) {
-      Map<String, String> parameters = parseParameters(query);
-      String region;
-      if (parameters.containsKey(REGION)) {
-        region = parameters.get(REGION);
-      } else {
+      LOGGER.debug("Building new S3Client for query '{}'.", query);
+      Map<String, List<String>> parameters;
```

Comment:
I wonder if these checks shouldn't be placed outside of this method, and before we try to store a client in the cache?

Reply:
This ties into your other comment about semantically-equivalent queries.

On the one hand, parsing and validating the parameters on every URL is going to be much more expensive than only checking when we find a set of credentials without a cached client. This is especially true since I don't really expect people to be providing a large number of different sets of credentials, so it will mostly be parsing the same query over and over again.

On the other hand, you never know what your users might do! :laughing: So I'll go with the pre-parse-and-`S3ClientInfo`-key approach. We could potentially add _another_ cache for raw-query-to-`S3ClientInfo`, but that seems a bit like overkill to me. Let me know if you'd prefer a different idea.

Also, sorry for disappearing for so long; things have been pretty busy on my end. Hopefully I'll be able to squeeze in enough time to finish things up here. before too much longer :)